Test Upcoming Movies Feature - Usecase

Belal-Alkhatib opened the pull request:
# Pull Request Template

## Description
This PR introduces a comprehensive test suite for the GetUpcomingMoviesUseCase class and for HomeViewModel.
---

## Changes Made
List the changes introduced in this pull request:
**Added GetUpcomingMoviesUseCaseTest.kt with the following test cases:**

- All genres: Verifies all movies are returned when MovieGenre.ALL is passed.
- Genre filtering: Ensures only movies matching the provided genre are returned.
- No match by genre: Validates that an empty list is returned when no movies match the specified genre.
- Empty repository response: Ensures an empty list is returned when the repository provides no movies.
- Empty categories filtering: Confirms that movies with empty categories are excluded when filtering by genre.
- Exception propagation: Ensures repository exceptions (e.g., NoInternetException) are not swallowed.
- Sorting logic: Verifies sorting behavior: movies are sorted first by popularity (desc), then rating (desc).

**Added HomeViewModelTest.kt for upcoming movies feature with the following test cases:**
- Initial data load
- Genre selection
- Navigation triggers
- Error handling

---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

Belal-Alkhatib committed:
fix: ensure movies have categories before filtering

The `GetUpcomingMoviesUseCase` now checks if a movie has any categories before attempting to filter by a specific genre. This prevents potential errors if a movie's category list is empty.

Additionally, this commit introduces unit tests for `GetUpcomingMoviesUseCase` to cover various scenarios, including different genre filters, empty repository responses, and exception handling.

Belal-Alkhatib committed:
refactor: sort upcoming movies by popularity and rating

Belal-Alkhatib committed:
fix: app bar visibility

ThoriaHamdy99 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/upcoming-movies-feature

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt

eslam-99 commented on ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt:
```diff
@@ -136,7 +137,8 @@ private fun HomeScreenContent(
                     .statusBarsPadding()
-                    .padding(horizontal = 16.dp, vertical = 8.dp),
+                    .padding(horizontal = 16.dp, vertical = 8.dp)
+                    .zIndex(4f),
```
@Noor-Serry already fixed `zIndex` issue

Belal-Alkhatib committed:
refactor: remove zIndex from HomeAppBar

Belal-Alkhatib committed:
test: add HomeViewModel tests and refine upcoming movies logic

Belal-Alkhatib committed:
Merge remote-tracking branch 'origin/test/upcoming-movies-feature' into test/upcoming-movies-feature

# Conflicts:
#	ui/src/main/java/com/example/ui/screens/home/HomeScreen.kt
#	viewModel/src/main/java/com/example/viewmodel/home/HomeUiStateMapper.kt
#	viewModel/src/main/java/com/example/viewmodel/home/HomeViewModel.kt

Belal-Alkhatib committed:
refactor: update HomeViewModel to use GetHomeScreenDataUseCase

Belal-Alkhatib committed:
Merge branch 'develop' into test/upcoming-movies-feature

# Conflicts:
#	domain/src/test/java/com/example/domain/useCase/GetUpcomingMoviesUseCaseTest.kt

Belal-Alkhatib committed:
refactor: remove zIndex from HomeTopAppBar

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/shared/BaseViewModel.kt:
```diff
@@ -27,7 +27,7 @@ open class BaseViewModel<S, E>(
 
 
-    protected fun updateState(updater: (S) -> S) {
+    fun updateState(updater: (S) -> S) {
```
This change is for testing only, I see the solution for that is calling public functions in viewmodel that updates state instead of updating state manually
I tried this before in `SearchByCountryViewModel` and got 100% coverage so it should be possible here too

Noor-Serry replied:
actually he dose not need to change the it

Belal-Alkhatib replied:
ok i will do that 👍

Noor-Serry commented on viewModel/src/test/java/com/example/viewmodel/home/HomeViewModelTest.kt:
```diff
@@ -0,0 +1,240 @@
+        coEvery { getUpcomingMoviesUseCase(any()) } returns upcomingMovies
+        every { homeUiStateMapper.moviesToMoviesItemsUiState(upcomingMovies) } returns expectedUiState
+
+        viewModel.updateState { it.copy(error = HomeError.NetworkError) }
```
that is wrong test we should test the state not update the state from the test function 
you should test that current state match the expected state after preform specific action(method)

Belal-Alkhatib replied:
done

Belal-Alkhatib committed:
test: enhance retry logic and test cases in HomeViewModelTest

This commit enhances the test cases for the `onClickRetryLoading` function in `HomeViewModelTest` to cover scenarios where a network error occurs during the initial load attempt.

The following changes were made:
- Added test cases to verify that `onClickRetryLoading` clears the error state and reloads movies even when a network error occurs.
- Modified the visibility of the `updateState` function in `BaseViewModel` from public to protected.

Noor-Serry commented on viewModel/src/test/java/com/example/viewmodel/home/HomeViewModelTest.kt:
```diff
@@ -0,0 +1,254 @@
+            viewModel.onChangeUpcomingMovieGenre(genre)
+            advanceUntilIdle()
+
+            coVerify(exactly = 0) { getUpcomingMoviesUseCase(any()) }
```
you just assert the state and the effect when you test the view model so here you should test that the state its not changed

Belal-Alkhatib committed:
test: ensure onChangeUpcomingMovieGenre doesn't recall when genre hasn't changed
